kamu-adapter-http/src/simple_protocol/handlers.rs

```diff
@@ -154,6 +164,81 @@ pub async fn dataset_checkpoints_handler(
 
 /////////////////////////////////////////////////////////////////////////////////
 
+pub async fn dataset_data_put_handler(
+    catalog: Extension<dill::Catalog>,
+    axum::extract::Path(hash_param): axum::extract::Path<PhysicalHashFromPath>,
+    axum::extract::Query(staging_name_param): axum::extract::Query<StagingNameFromQuery>,
+    mut body_stream: axum::extract::BodyStream,
+) -> Result<(), axum::http::StatusCode> {
+    dataset_put_object_common(
+        catalog,
+        hash_param.physical_hash,
+        staging_name_param.staging_name.as_str(),
+        "data",
+        &mut body_stream,
+    )
+    .await
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+pub async fn dataset_checkpoints_put_handler(
+    catalog: Extension<dill::Catalog>,
+    axum::extract::Path(hash_param): axum::extract::Path<PhysicalHashFromPath>,
+    axum::extract::Query(staging_name_param): axum::extract::Query<StagingNameFromQuery>,
+    mut body_stream: axum::extract::BodyStream,
+) -> Result<(), axum::http::StatusCode> {
+    dataset_put_object_common(
+        catalog,
+        hash_param.physical_hash,
+        staging_name_param.staging_name.as_str(),
+        "checkpoints",
+        &mut body_stream,
+    )
+    .await
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+async fn dataset_put_object_common(
+    catalog: Extension<dill::Catalog>,
+    physical_hash: Multihash,
+    staging_name: &str,
+    internal_folder_name: &str,
+    body_stream: &mut axum::extract::BodyStream,
+) -> Result<(), axum::http::StatusCode> {
+    let workspace_layout = catalog
+        .get_one::<WorkspaceLayout>()
+        .map_err(|_| axum::http::StatusCode::INTERNAL_SERVER_ERROR)?;
+
+    let temp_data_folder_suffix = format!(
+        "{}/{}/{}",
+        staging_name,
+        internal_folder_name,
+        get_staging_name()
+    );
+
+    let temp_upload_path = workspace_layout.datasets_dir.join(temp_data_folder_suffix);
+    upload_file_from_axum_to_disk(&temp_upload_path, body_stream)
```

Comment:
Can't we stream object in via `ObjectRepository` interface, so this is generic over whatever storage implementation we chose (lfs, in-memory, proxied S3)?

Reply:
We can't. We do not have any way to obtain the correct `ObjectRepository` in the 2ndary call that has no access to the `DatasetBuilder`

Reply:
Resolved